sequencer/src/block2/tables.rs

```diff
@@ -0,0 +1,260 @@
+use crate::{BlockBuildingSnafu, Error, VmId};
+use derivative::Derivative;
+use serde::{Deserialize, Serialize};
+use snafu::OptionExt;
+use std::marker::PhantomData;
+use std::mem::size_of;
+use std::ops::Range;
+
+#[derive(Clone, Debug, Derivative, Deserialize, Eq, Serialize)]
+#[derivative(Hash, PartialEq)]
+
+// TODO (Philippe) make it private?
+pub struct NamespaceInfo {
+    // `tx_table` is a bytes representation of the following table:
+    // word[0]: [number n of entries in tx table]
+    // word[j>0]: [end byte index of the (j-1)th tx in the payload]
+    //
+    // Thus, the ith tx payload bytes range is word[i-1]..word[i].
+    // Edge case: tx_table[-1] is implicitly 0.
+    //
+    // Word type is `TxTableEntry`.
+    //
+    // TODO final entry should be implicit:
+    // https://github.com/EspressoSystems/espresso-sequencer/issues/757
+    pub(crate) tx_table: Vec<u8>,
+    pub(crate) tx_bodies: Vec<u8>, // concatenation of all tx payloads
+
+    #[derivative(Hash = "ignore")]
+    #[derivative(PartialEq = "ignore")]
+    #[serde(skip)]
+    pub(crate) tx_bytes_end: TxTableEntry,
+    pub(crate) tx_table_len: TxTableEntry,
+}
+
+pub trait Table<TableWord: TableWordTraits> {
+    // Read TxTableEntry::byte_len() bytes from `table_bytes` starting at `offset`.
+    // if `table_bytes` has too few bytes at this `offset` then pad with zero.
+    // Parse these bytes into a `TxTableEntry` and return.
+    // Returns raw bytes, no checking for large values
+    fn get_table_len(&self, offset: usize) -> TxTableEntry;
+
+    fn get_payload(&self) -> Vec<u8>;
+
+    fn byte_len() -> usize {
+        size_of::<TableWord>()
+    }
+}
+
+impl<TableWord: TableWordTraits> Table<TableWord> for NameSpaceTable<TableWord> {
+    // TODO (Philippe) avoid code duplication with similar function in TxTable?
+    fn get_table_len(&self, offset: usize) -> TxTableEntry {
+        let end = std::cmp::min(
+            offset.saturating_add(TxTableEntry::byte_len()),
+            self.raw_payload.len(),
+        );
+        let start = std::cmp::min(offset, end);
+        let tx_table_len_range = start..end;
+        let mut entry_bytes = [0u8; TxTableEntry::byte_len()];
+        entry_bytes[..tx_table_len_range.len()]
+            .copy_from_slice(&self.raw_payload[tx_table_len_range]);
+        TxTableEntry::from_bytes_array(entry_bytes)
+    }
+
+    fn get_payload(&self) -> Vec<u8> {
+        self.raw_payload.clone()
+    }
+}
+
+impl<TableWord: TableWordTraits> NameSpaceTable<TableWord> {
+    pub fn from_vec(v: Vec<u8>) -> Self {
+        Self {
+            raw_payload: v,
+            phantom: Default::default(),
+        }
+    }
+
+    pub fn from_bytes(b: &[u8]) -> Self {
+        Self {
+            raw_payload: b.to_vec(),
+            phantom: Default::default(),
+        }
+    }
+
+    pub fn get_bytes(&self) -> Vec<u8> {
+        self.raw_payload.clone()
+    }
+
+    pub fn add_new_entry_vmid(&mut self, id: VmId) -> Result<(), Error> {
```

Comment:
Regarding `add_new_entry_vmid` and `add_new_entry_payload_len`: similar comment as above regarding `Payload::{generate_raw_payload, update_namespace_with_tx}`. These are helper methods used only in construction. They should not be `pub`. Instead their scope should be severely restricted so that no one can accidentally call them.

Reply:
Right, done in c88bd03.